packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/gridRowGroupingUtils.ts

```diff
@@ -76,22 +76,29 @@ export const filterRowTreeFromGroupingColumns = (
     areAncestorsExpanded: boolean,
   ): number => {
     let isMatchingFilters: boolean;
+    let isMatchingFilterItems: boolean;
+    let isMatchingQuickFilter: boolean;
     if (!isRowMatchingFilters) {
+      isMatchingFilterItems = true;
+      isMatchingQuickFilter = true;
       isMatchingFilters = true;
     } else {
       const shouldApplyItem = node.isAutoGenerated
         ? (columnField: string) => shouldApplyFilterItemOnGroup(columnField, node)
         : undefined;
 
-      isMatchingFilters = isRowMatchingFilters(node.id, shouldApplyItem);
+      const { passFilterItems, passQuickFilter } = isRowMatchingFilters(node.id, shouldApplyItem);
+      isMatchingFilterItems = passFilterItems;
+      isMatchingQuickFilter = passQuickFilter;
+      isMatchingFilters = isMatchingFilterItems && isMatchingQuickFilter;
     }
 
     let filteredDescendantCount = 0;
     node.children?.forEach((childId) => {
       const childNode = rowTree[childId];
       const childSubTreeSize = filterTreeNode(
         childNode,
-        areAncestorsPassingChildren && isMatchingFilters,
+        areAncestorsPassingChildren && isMatchingFilterItems,
```

Comment:
Maybe that could fix the problem of OR operator

```suggestion
        areAncestorsPassingChildren && (isMatchingFilterItems || filterModel.linkOperator==="Or"),
```

Reply:
I find this part problematic:

https://github.com/mui/mui-x/blob/4dabd10c9b264ea83e8425a168e7086985bf242d/packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx#L123-L126

I believe it causes this issue: https://github.com/mui/mui-x/pull/5314#issuecomment-1183331537
Basically if you have multiple filters where one of them filters by grouping column **and** have `OR` link operator - then all the rows will pass the filters, because every real row will fall into the `if` statement above and therefore each row will always have at least one filter passing. And since the link operator is `OR` - it's enough for row to pass all the filters.

Shouldn't we return `true` if `linkOperator === GridLinkOperator.And` and `false` if `linkOperator === GridLinkOperator.Or`

Reply:
Then we would also need this change:

```diff
-let isPassingFiltering = isMatchingFilters;
+let isPassingFiltering: boolean;
+if (linkOperator === GridLinkOperator.And) {
+  isPassingFiltering = isMatchingFilters && areAncestorsPassingChildren;
+} else if (linkOperator === GridLinkOperator.Or) {
+  isPassingFiltering = isMatchingFilters || areAncestorsPassingChildren;
+}

Reply:
Yes, for now tests were passing because I used the column field instead of the grouping field in filter model

Reply:
It seems we reached the limit of having a single function that for a row says if it passes the filtering or not.

If we do the modification you propose, we make a `getApplyFilterFn` dependent on the `linkOperator` which is possible for this specific use case but will not scale on more complex ones like filter with and/or combinations:

> to pass row should validate (filterItemA and filterItemB) or (filterItemC and filterItemD)

A solution could be to refactor the behavior of the helper function. Instead of returning a boolean if the row passes the filter or not, it would return for each item if it's passing or not.

So in a tree structure, we could stack the results of all the parents and then take the decision according to the linkOperator